### PR TITLE
Adds closeLine and withdraw functions, and corresponding tests

### PR DESCRIPTION
**Pull request by denizmatar**

### CODE REVIEW

File: contracts/defter.sol [hunk 1]
```diff
@@ -2,6 +2,7 @@
 pragma solidity ^0.8.0;
 
 import "hardhat/console.sol";
+import "contracts/IERC20.sol";
```
Comment (randabat) [inline contracts/defter.sol]: import from open-zeppelin

File: contracts/defter.sol [hunk 2]
```diff
@@ -25,19 +26,36 @@ contract Defter {
     */
     mapping(bytes32 => mapping(address => uint256)) balances;
 
-    event OpenLine(
+    event LineOpened(
```
Comment (randabat) [inline contracts/defter.sol:37]: emit single event for each receiver with `indexed` key

File: contracts/defter.sol [hunk 3]
```diff
@@ -25,19 +26,36 @@ contract Defter {
...
         address indexed from,
-        // address[] receivers,
-        // uint256[] amounts,
+        address[] receivers,
+        uint256[] amounts,
         bytes32 indexed lineID
     );
 
-    event TransferLine(address indexed from, bytes32 indexed lineID);
+    event LineTransferred(
```
Comment (randabat) [inline contracts/defter.sol:44]: emit single event for each receiver with `indexed` key

File: contracts/defter.sol [hunk 4]
```diff
@@ -50,43 +68,114 @@ contract Defter {
...
-        // check maturity
-        require(_maturityDate > (block.timestamp + 1 days));
+    ) external {
+        openLineHelper(_maturityDate, _unit, receivers, amounts);
+    }
+
+    function openLineHelper(
+        uint256 _maturityDate,
+        address _unit,
+        address[] memory receivers,
```
Comment (randabat) [inline contracts/defter.sol]: take single receiver in this function. iterate receivers above

File: contracts/defter.sol [hunk 5]
```diff
@@ -50,43 +68,114 @@ contract Defter {
...
+    //     IERC20(tokenContractAddress).transferFrom(sender, recipient, amount);
+
+    //     return true;
+    // }
+
     function openLine(
         uint256 _maturityDate,
-        string memory _unit,
+        address _unit,
         address[] memory receivers,
```
Comment (randabat) [inline contracts/defter.sol]: check lenght of `receivers` and `amounts` match

File: contracts/defter.sol (hunk 4 above)
Comment (randabat) [reply contracts/defter.sol]: disallow 0 length

File: contracts/defter.sol [hunk 6]
```diff
@@ -50,43 +68,114 @@ contract Defter {
...
         }
-
-        // madem alicilari da log'da saklayalim?
-        emit OpenLine(msg.sender, _lineID);
+        emit LineOpened(msg.sender, receivers, amounts, _lineID);
     }
 
     function transferLine(
         bytes32 lineID,
         address[] memory receivers,
```
Comment (randabat) [inline contracts/defter.sol]: same as above

File: contracts/defter.sol [hunk 7]
```diff
@@ -50,43 +68,114 @@ contract Defter {
...
+        // illa sartsa bir cozum uretebiliriz
+        emit LineTransferred(msg.sender, receivers, amounts, lineID);
+    }
+
+    function closeLine(
+        uint256 _maturityDate,
+        address _unit,
+        uint256 _totalAmount
+    ) public {
+        closeLineHelper(_maturityDate, _unit, _totalAmount);
```
Comment (randabat) [inline contracts/defter.sol]: helper is not required here since it's single operation

File: contracts/defter.sol [hunk 8]
```diff
@@ -50,43 +68,114 @@ contract Defter {
...
+        // emitleri loop'un icine alamadim cunku testini yazamadim o sekilde
+        // illa sartsa bir cozum uretebiliriz
+        emit LineTransferred(msg.sender, receivers, amounts, lineID);
+    }
+
+    function closeLine(
+        uint256 _maturityDate,
+        address _unit,
+        uint256 _totalAmount
+    ) public {
```
Comment (randabat) [inline contracts/defter.sol]: `external` key is better here

File: contracts/defter.sol [hunk 9]
```diff
@@ -50,43 +68,114 @@ contract Defter {
...
+        IERC20 token = IERC20(_unit);
+        bytes32 _lineID = hashLine(msg.sender, _maturityDate, _unit);
+        token.transferFrom(msg.sender, address(this), _totalAmount);
+
+        emit LineClosed(msg.sender, _lineID, _totalAmount);
+        // artik withdraw edilebildigini belirtmem lazim ama nasil?
+        // front end emit'i kontrol ederek bu isi halledebilir mi?
+    }
+
+    function withdraw(bytes32 _lineID, address _unit) public {
```
Comment (randabat) [inline contracts/defter.sol]: should use `external`

File: contracts/defter.sol [hunk 10]
```diff
@@ -50,43 +68,114 @@ contract Defter {
...
+        closeLineHelper(_maturityDate, _unit, _totalAmount);
+    }
+
+    // total amount parametre olarak alsin, ve front'end hesaplayip gondersin.
+    // cunku burda for loop yapmak zorundayim, gereksiz maliyet gibi geldi.
+    // balances[_linedID][0] yazmissin ama address bekledigini soyluyor bana
+    function closeLineHelper(
+        uint256 _maturityDate,
+        address _unit,
+        uint256 _totalAmount
```
Comment (randabat) [inline contracts/defter.sol]: we should check if totalAmount actually holds. So then let's create a `Line` struct with `totalAmount` field and check it here.

File: contracts/defter.sol (hunk 9 above)
Comment (randabat) [reply contracts/defter.sol]: line status must be closed it can be done with a status field in Line struct

**Review by randabat — CHANGES_REQUESTED** [review]

File: contracts/defter.sol (hunk 10 above)
Comment (denizmatar) [reply contracts/defter.sol]: done

File: contracts/defter.sol (hunk 9 above)
Comment (denizmatar) [reply contracts/defter.sol]: external done
changed line status at the end of closeLine instead of withdraw

File: contracts/defter.sol (hunk 8 above)
Comment (denizmatar) [reply contracts/defter.sol]: done

File: contracts/defter.sol (hunk 7 above)
Comment (denizmatar) [reply contracts/defter.sol]: done

File: contracts/defter.sol (hunk 6 above)
Comment (denizmatar) [reply contracts/defter.sol]: done

File: contracts/defter.sol (hunk 5 above)
Comment (denizmatar) [reply contracts/defter.sol]: done

File: contracts/defter.sol (hunk 4 above)
Comment (denizmatar) [reply contracts/defter.sol]: done

File: contracts/defter.sol (hunk 3 above)
Comment (denizmatar) [reply contracts/defter.sol:44]: done

File: contracts/defter.sol (hunk 2 above)
Comment (denizmatar) [reply contracts/defter.sol:37]: done

File: contracts/defter.sol (hunk 1 above)
Comment (denizmatar) [reply contracts/defter.sol]: done

File: test/defter_test.ts [hunk 11]
```diff
@@ -94,14 +94,13 @@ describe("Defterhane", () => {
...
                 ),
-            )
-                .to.emit(defter, "LineOpened")
-                .withArgs(
-                    owner.address,
-                    [addr1.address, addr2.address],
-                    [20, 30],
-                    hashedLine,
-                )
+            ).to.emit(defter, "LineOpened")
```
Comment (randabat) [inline test/defter_test.ts]:
```
                .to.emit(defter, "LineOpened")
                .withArgs(owner.address, addr1.address, 20, hashedLine)
                .to.emit(defter, "LineOpened")
                .withArgs(owner.address, addr2.address, 30, 
```

File: contracts/defter.sol [hunk 12]
```diff
@@ -177,5 +184,7 @@ contract Defter {
         }
     }
 
+    receive() external payable {}
```
Comment (randabat) [inline contracts/defter.sol]: receive and fallback functions are nod required

File: test/defter_test.ts [hunk 13]
```diff
@@ -173,9 +172,8 @@ describe("Defterhane", () => {
                 defter
                     .connect(addr2)
                     .transferLine(hashedLine, [addr1.address], [30]),
-            )
-                .to.emit(defter, "LineTransferred")
-                .withArgs(addr2.address, [addr1.address], [30], hashedLine)
+            ).to.emit(defter, "LineTransferred")
```
Comment (randabat) [inline test/defter_test.ts]: same as above

**Review by randabat — CHANGES_REQUESTED** [review]

File: test/defter_test.ts (hunk 13 above)
Comment (denizmatar) [reply test/defter_test.ts]: done

File: contracts/defter.sol (hunk 12 above)
Comment (denizmatar) [reply contracts/defter.sol]: done

File: test/defter_test.ts (hunk 11 above)
Comment (denizmatar) [reply test/defter_test.ts]: done

**Pull request merged**